docs/plutus-ir/plutus-ir.md

```diff
@@ -0,0 +1,193 @@
+number of names with various types or kinds, with the correctness conditions being:
+- The type variable declarations are well-kinded.
+- The application of the datatype to its type variables is well-kinded and of kind `(type)`.
+- The constructor argument types are all well-kinded and of type `(type)`.
+
+## Design notes
+
+### Why do we need variadic syntax?
+
+`letrec` is inherently about binding an arbitrary number of mutually recursive bindings, so
+we need to be able to write that.
+
+### Why is `let` variadic?
+
+`letrec` obviously needs to be variadic, since the whole point is to have multiple bindings
+that reference each other.
+
+`let` doesn't *need* to be variadic, but it's symmetrical, convenient, and no harder to support.
+
+### Why is there a special datatype binding?
+
+Datatype bindings are weird in that they bind multiple names, which are part of the binding
+(the type, the constructors, the destructor). So we can't handle it as a normal "type binding" since
+it also declares other things.
+
+### Why are variable declarations reified in the syntax?
+
+It's very helpful to have a syntax form for variable declarations, since the syntax is too
+homogeneous when everything is just names. For example, in the datatype declaration we have
+a whole series of type variable declarations. I think we could just about parse the version
+with them all flattened, but it's much clearer when they're reified.
+
+I think this is quite useful and arguably worth introducing in the Plutus Core syntax too.
+
+### Why does `datatype` bind its type variables?
+
+I considered having the constructors of the datatype have an explicit type instead of having arguments, e.g.
+`(constr Cons (all a (type) (fun a (fun [List a] [List a]))))`.
+
+The downside of this is that:
+- It obscures the nature of the datatype as a sum-of-products.
+- We need to construct several types from the constructor: the main datatype, the matcher type, and the constructor
+  types. Having the "full" constructor type given means that we just need to deconstruct it again later.
+
+Plus it's even closer to the usual Haskell-style `data` declarations.
+
+## Discussion points
+
+### I hate your keyword suggestions
+
+Better suggestions welcome! In particular I wonder whether we can get away with using `con` instead of `constr`.
+(I believe in an older version of the Plutus Core spec which had datatypes, `con` was indeed used for constructors.)
+
+### Should we have `match`?
+
+Currently there is no `match` syntax, instead we expose destructor functions.
+
+This has the following advantages:
+- Easier to translate into Plutus Core
+- Only allowing "simple" matches, which are all we're going to support anyway.
+
+It does have some disadvantages:
+- Need an additional name for the destructor.
+- The destructor needs to be instantiated at the result type of the match, whereas for a match
+  syntax this could be inferred.
+
+I think these are fine for an IR.
```

Comment:
I agree.

Reply:
I'm not sure what "explicit destructors" means in the above.

I think the IR should hide the destructor functions, and that we should have match instead, and that match should support nested patterns.

Reason one is to consolidate an important operation into the IR. We want the IR to help us compile from Plutus Tx. Presumably, Plutus Tx will allow pattern matching in equations. Compiling nested patterns in the IR will make it easier to compile Plutus Tx to the IR, and will make it easy for others to use nested patterns. This matches closely our motivation for introducing an IR to begin with.

Reason two is aesthetic. Ever since Gentzen, we design our logics and our programming languages by introducing constructors and destructors for each type. For a datatype, the constructors are the constructors (duh), and the destructor is match. I think it's bad taste to have a special form for the constructor, but just treat the destructor as a function.

Reply:
In practical terms, adding `match` to Plutus IR won't help us much with the *current* GHC Core compiler. By the time we reach GHC Core most of the complicated pattern matching stuff has been compiled away. In particular, nested patterns do not appear.

I don't understand your second point. We don't have a special form for the constructor. Both the constructors and destructors are just functions. This seems more symmetrical to me than having a special form for the destructor (match syntax) and not for contructors.